File: src/main/java/com/gmail/borlandlp/minigamesdtools/arena/team/TeamController.java
package com.gmail.borlandlp.minigamesdtools.arena.team;

import com.gmail.borlandlp.minigamesdtools.arena.ArenaBase;
import com.gmail.borlandlp.minigamesdtools.arena.ArenaPhaseComponent;
import com.gmail.borlandlp.minigamesdtools.arena.ArenaPlayersRelative;
import com.gmail.borlandlp.minigamesdtools.events.ArenaPlayerQuitEvent;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.scoreboard.Team;

import java.util.*;

public class TeamController implements ArenaPhaseComponent {
	private ArrayList<TeamProvider> teams = new ArrayList<>();
	private ExampleTeamListener exampleTeamListener;
	private ArenaBase arena;

	public TeamController(ArenaBase arena) {
		this.arena = arena;
	}

	public ArenaBase getArena() {
		return arena;
	}

	public void addTeam(TeamProvider team) {
		this.teams.add(team);
	}

	public void removeTeam(TeamProvider team) {
		this.teams.remove(team);
	}

	public ArrayList<TeamProvider> getTeams() {
		return this.teams;
	}

	@Override
	public void onInit() {
		this.exampleTeamListener = new ExampleTeamListener(this);
		try {
			this.getArena().getEventAnnouncer().register(this.exampleTeamListener);
		} catch (Exception e) {
			e.printStackTrace();
		}


		for(TeamProvider team : this.getTeams()) {
			this.getArena().getPhaseComponentController().register(team);
		}
	}

    @Override
	public void beforeGameStarting() {

	}

	@Override
	public void gameEnded() {
		this.getArena().getEventAnnouncer().unregister(this.exampleTeamListener);
		ArenaPlayerQuitEvent event;
		for (TeamProvider team : this.getTeams()) {
			for (Player player : Objects.requireNonNull(team.getPlayers())) {
				//team.removePlayer(player);
				event = new ArenaPlayerQuitEvent(this.getArena(), player);
				Bukkit.getPluginManager().callEvent(event);
			}
		}
	}

	@Override
	public void update() {

	}

	@Override
	public void beforeRoundStarting() {

	}

	@Override
	public void onRoundEnd() {

	}

	public int countCurrentPlayers() {
		return this.getTeams().stream().mapToInt(t -> t.getPlayers().size()).sum();
	}

	public TeamProvider getTeam(String teamName) {
		return this.getTeams().stream()
				.filter(team -> team.getName().equals(teamName))
				.findFirst()
				.orElse(null);
	}

	public TeamProvider getTeamOf(String nickname) {
		for(TeamProvider team : this.getTeams()) {
			for (Player player : team.getPlayers()) {
				if(player.getName().equals(nickname)) return team;
			}
		}

		return null;
	}

	public TeamProvider getTeamOf(Player player) {
		return this.getTeamOf(player.getName());
	}

	public ArenaPlayersRelative getPlayersRelative(Player player1, Player player2) {
		TeamProvider team1 = this.getTeamOf(player1);
		TeamProvider team2 = this.getTeamOf(player2);

		if(team1 == null || team2 == null) {
			return ArenaPlayersRelative.ENEMY;
		} else if(team1 == team2 && !team1.friendlyFireAllowed()) {
			return ArenaPlayersRelative.TEAMMATE;
		} else {
			return ArenaPlayersRelative.ENEMY;
		}
	}
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/arena/team/TeamController.kt
package com.gmail.borlandlp.minigamesdtools.arena.team

import com.gmail.borlandlp.minigamesdtools.arena.ArenaBase
import com.gmail.borlandlp.minigamesdtools.arena.ArenaEventListener
import com.gmail.borlandlp.minigamesdtools.arena.ArenaPhaseComponent
import com.gmail.borlandlp.minigamesdtools.arena.ArenaPlayersRelative
import com.gmail.borlandlp.minigamesdtools.events.ArenaPlayerQuitEvent
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import java.util.*

class TeamController(val arena: ArenaBase) : ArenaPhaseComponent {
    val teams = ArrayList<TeamProvider>()
    private var exampleTeamListener: ExampleTeamListener? = null

    fun addTeam(team: TeamProvider) {
        teams.add(team)
    }

    fun removeTeam(team: TeamProvider?) {
        teams.remove(team)
    }

    override fun onInit() {
        exampleTeamListener = ExampleTeamListener(this)
        try {
            arena.eventAnnouncer.register(exampleTeamListener!!)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        for (team in teams) {
            arena.phaseComponentController.register(team)
        }
    }

    override fun beforeGameStarting() {}
    override fun gameEnded() {
        arena.eventAnnouncer.unregister(exampleTeamListener as ArenaEventListener)
        var event: ArenaPlayerQuitEvent
        for (team in teams) {
            for (player in team.players) { //team.removePlayer(player);
                event = ArenaPlayerQuitEvent(arena, player)
                Bukkit.getPluginManager().callEvent(event)
            }
        }
    }

    override fun update() {}
    override fun beforeRoundStarting() {}
    override fun onRoundEnd() {}
    fun countCurrentPlayers(): Int {
        return teams.stream().mapToInt { t: TeamProvider -> t.players.size }
            .sum()
    }

    fun getTeam(teamName: String?): TeamProvider? {
        return teams.stream()
            .filter { team: TeamProvider? -> team!!.name.equals(teamName) }
            .findFirst()
            .orElse(null)
    }

    fun getTeamOf(nickname: String): TeamProvider? {
        for (team in teams) {
            for (player in team.players) {
                if (player.name == nickname) return team
            }
        }
        return null
    }

    fun getTeamOf(player: Player): TeamProvider? {
        return this.getTeamOf(player.name)
    }

    fun getPlayersRelative(player1: Player, player2: Player): ArenaPlayersRelative {
        val team1 = this.getTeamOf(player1)
        val team2 = this.getTeamOf(player2)
        return if (team1 == null || team2 == null) {
            ArenaPlayersRelative.ENEMY
        } else if (team1 === team2 && !team1.isFriendlyFireAllowed) {
            ArenaPlayersRelative.TEAMMATE
        } else {
            ArenaPlayersRelative.ENEMY
        }
    }

}